[Jonathan Yeo] iP

JonathanHoshi opened the pull request:
## JJBA
JJBA helps you record things you need to do so you wont have to remember it. It
- is Text-based
- is EASY to learn
- is FAST to use
- ~~was me, DIO!~~
![image](https://user-images.githubusercontent.com/28040511/151690377-8c84de02-a021-405e-9ca6-e8ee85fecee1.png)
> "KONO DIO DA!" - Dio Brando [(source)](https://gfascii.art/2019/08/dio-brando/)

All you need to do is,
1. Download from [here](https://github.com/JonathanHoshi/ip).
2. Run it.
3. Add you task.
4. Let it keep track of your task for you. 😁
5. ~~Find DIO to let him dominate the bot! 🔪~~

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] MUDA MUDA MUDA MUDA MUDA
- [ ] Reminders

---
If you are trained as a ~~Hamon user~~ Java programmer, you can use it to practice ~~Hamon~~ Java too. Here's the `main` method:

```java
public static void main(String[] args) {
    new Duke(FILE_PATH, new JJBABotMessage()).run();    //FILE_PATH = "/data/taskInfo.txt";
}
```

Commit:
Add Gradle support

JonathanHoshi committed:
Added Greet, Echo and Exit functionality

console formatting class and bot command functionality

JonathanHoshi committed:
Added Add and List functionality

store a list of the user inputs and list all inputs

JonathanHoshi committed:
Added Mark, Unmark functionality

added ability to mark and unmark task and option to switch to dio bot

JonathanHoshi committed:
Added Todo, Event and Deadline functionality

added handling of different type of task, modified structure of task accommodate for the different tasks

JonathanHoshi committed:
Tested UI Testing and fixed todo

fixed todo printing wrong output

JonathanHoshi committed:
Added error handling.

implement 'mark' invalid input and out of range exception, added overall empty input exception and invalid suffix exception.

JonathanHoshi committed:
Added Delete feature, and further testing

implement delete feature and update error message to show warning symbol and restructure the printing of task left to be generic for usage in deletion. Testing of the new functions and ensure correct output

JonathanHoshi committed:
Organize and Restructure error checking code

move error checking into individual methods for organization

JonathanHoshi committed:
Add Save and Load functionality

implement storage class which saves and load the tasklist.

JonathanHoshi committed:
Update variable names

change variable name to make more sense

JonathanHoshi committed:
Add Date/Time support to tasks

implement deadline and event to work with localdate and localtime.

JonathanHoshi committed:
Merge branch 'branch-Level-7'

JonathanHoshi committed:
Merge branch 'branch-Level-8'

JonathanHoshi committed:
Restructure classes to become more OOP

Duke class: extract out console, command & parser class.

Extracting the different classes allow each class to handle their individual functionality. This allows for easier troubleshooting and re-use of code (inheritance).

Storage, Console, Command, DukeException classes is affected by OOP restructuring. Output of commands is encapsulated in a class 'BotMessage' and Command word and format is encapsulated in their individual subclasses of Command.

Other minor changes: clean up code spacing.

JonathanHoshi committed:
Organize classes into packages

Divide all classes into multiple packages due to the increase number of classes.

JonathanHoshi committed:
Fix Parser class formatting issue

Parser class methods did not use the constants provided.

Make methods use constants to make it easier for user to change date time format.

JonathanHoshi committed:
Add JUnit testing and Fix name merge issues

Require testing for classes to ensure proper behaviour of code.

Implement ParserTest,  DukeTest and DeadlineTest to test functionality of Parser, Duke and Deadline classes.

Other info
- Add a static formatMessage method to Console.
- Extract Duke commandFeedback handling into a method.
- Add a method to all commands to get command name.
- Fix naming issues in BotMessage and Add missing @Override tag in Task

JonathanHoshi committed:
Fix Storage loading wrong file path and minor issues

Storage class was loading the wrong file path, /data/taskInfo instead of <user.dir>/data/taskInfo. This results in Storage initializing creating the file at the wrong location compared to file read/write.

Other info
- Fix Console print issue (print instead of println).
- Fix JUnit testing to match with Console print.

JonathanHoshi committed:
Update code to comply with standard and minor bug fix

Split .* to individual imports and make storage class throw exception when directory and file cannot be created.

JonathanHoshi committed:
Add JavaDocs to all classes.

Add JavaDoc comments to the code.

JonathanHoshi committed:
Add Find command

Add a command to list the task that matches with the user input specified.

JonathanHoshi committed:
Merge branch 'branch-A-CodingStandard'

JonathanHoshi committed:
Merge branch 'branch-A-JavaDoc'

JonathanHoshi committed:
Merge branch 'branch-Level-9'

Resolve conflicts and Fix storage create folder/file bug.

JonathanHoshi committed:
Update JavaDoc for new Find command

Update the bot messages to have a javadoc header for the find methods.

JonathanHoshi committed:
Fix AJar and Add missing comment for FindCommand

Add missing parameter description for the FindCommand class

Add manifest file for AJar artifact build as shadow jar file was unable to run.

JonathanHoshi committed:
Fix build.gradle file to allow jar build to work

build.grade file path to the duke.java class was not correct.

JonathanHoshi committed:
Merge branch 'branch-A-Gradle'

JonathanHoshi committed:
Update code to match CheckStyle requirements

lots of stuff being updated in code to adhere to check style requirements

JonathanHoshi committed:
Merge branch 'branch-CheckStyle'

JonathanHoshi committed:
Add GUI Interface to Duke

Restructure UI (Console) and Duke class to work with new GUI implementation with FXML.

Features added
- GUI interface, text box for user input, scroll bar for looking through all messages, send button to send the input message to the bot.
- Each bot have their own image file

Other Info
- Rename Console class to UI class
- Move Task List printing to TaskList.toString().

Gnoznaug commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class FindCommand extends Command {
+    public static final String COMMAND_WORD = "find";
+    public static final String COMMAND_FORMAT = COMMAND_WORD + " <task description>";
```
Maybe can add a line break after the + operator.

Gnoznaug commented on src/main/java/duke/command/index/MarkCommand.java:
```diff
@@ -0,0 +1,84 @@
+ * depending on the status provided.
+ */
+public class MarkCommand extends IndexCommand {
+    public static final String COMMAND_WORD_MARK = "mark";
+    public static final String COMMAND_WORD_UNMARK = "unmark";
+    private static final String COMMAND_FORMAT_MARK = COMMAND_WORD_MARK + COMMAND_ARG;
+    private static final String COMMAND_FORMAT_UNMARK = COMMAND_WORD_UNMARK + COMMAND_ARG;
```
Followed constant name format. Very nice.

Gnoznaug commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+package duke.parser;
+
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+import duke.bot.BotType;
+import duke.command.BotCommand;
+import duke.command.Command;
+import duke.command.ExitCommand;
+import duke.command.FindCommand;
+import duke.command.ListCommand;
+import duke.command.add.DeadlineCommand;
+import duke.command.add.EventCommand;
+import duke.command.add.TodoCommand;
+import duke.command.index.DeleteCommand;
+import duke.command.index.MarkCommand;
+import duke.exception.DukeException;
+import duke.exception.InvalidArgumentException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Todo;
```
Followed the import convention well. Very nice.

Gnoznaug commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+     * @throws DukeException if command format is invalid.
+     */
+    public static Command parseCommand(String input) throws DukeException {
+        String[] inputArgs = input.trim().split(" ", 2);
```
Followed array specifier convention. Very nice.

Gnoznaug reviewed (commented):
LGTM! 👍

JonathanHoshi committed:
Merge branch 'branch-Level-10'

whoisjunhong commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,221 @@
+     * @return a command object.
+     * @throws DukeException if command format is invalid.
+     */
+    public static Command parseCommand(String input) throws DukeException {
+        String[] inputArgs = input.trim().split(" ", 2);
+
+        switch (inputArgs[0]) {
+        case TodoCommand.COMMAND_WORD:
+            return handleTodo(inputArgs);
+        case DeadlineCommand.COMMAND_WORD:
+            return handleDeadline(inputArgs);
+        case EventCommand.COMMAND_WORD:
+            return handleEvent(inputArgs);
+        case DeleteCommand.COMMAND_WORD:
+            return handleDelete(inputArgs);
+        case MarkCommand.COMMAND_WORD_MARK:
+            return handleMark(inputArgs, true);
+        case MarkCommand.COMMAND_WORD_UNMARK:
+            return handleMark(inputArgs, false);
+        case ListCommand.COMMAND_WORD:
+            return handleList(inputArgs);
+        case FindCommand.COMMAND_WORD:
+            return handleFind(inputArgs);
+        case ExitCommand.COMMAND_WORD:
+            return handleExit(inputArgs);
+        case BotCommand.COMMAND_WORD_JJBA:
+            return handleBot(inputArgs, BotType.JJBA);
+        case BotCommand.COMMAND_WORD_DIO:
+            return handleBot(inputArgs, BotType.DIO);
+        default:
+            throw new InvalidArgumentException();
+        }
```
I like how this is abstracted into a `Parser` class, this simplifies the rest of the other classes when parsing a duke command. 👍

whoisjunhong replied:
I like how this is abstracted into a Parser class, this simplifies the rest of the other classes when parsing a duke command. 👍

whoisjunhong commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,126 @@
+ * next instance of the program.
+ */
+public class Storage {
+    public static final String MESSAGE_READ_FAILURE = "Something went wrong with file read!";
+    public static final String MESSAGE_WRITE_FAILURE = "Something went wrong with file write!";
+    public static final String MESSAGE_INVALID_FILE = "File Corrupted!";
```
Good practice of static strings! 👍

whoisjunhong replied:
Good practice of static strings! 👍

whoisjunhong reviewed (changes requested):
Overall, your iP has the right amount of abstraction between the methods and classes.

I like how the JavaDoc is easy to read and understand what each method and class do. Great job! 👍

whoisjunhong reviewed (commented):
Overall, your iP has the right amount of abstraction between the methods and classes.

I like how the JavaDoc is easy to read and understand what each method and class do. Great job, LGTM! 👍

JonathanHoshi committed:
Format code to adhere code quality standards

Update JavaDocs of new methods in duke and package certain classes.

JonathanHoshi committed:
Merge pull request #1 from JonathanHoshi/branch-A-CodeQuality

Format code to adhere code quality standards

JonathanHoshi committed:
Add assertions

Allows to document important assumptions that should hold at certain points in the code.

This can be used to track the program state so that the runtime can verify these assumptions.

Add assert at: 
- Delete command to track size validity.
- Parser to track blank inputs.

JonathanHoshi committed:
Merge pull request #2 from JonathanHoshi/branch-A-Assertions

Add assertions

JonathanHoshi committed:
Update Exit command after GUI changes

Exit command was not updated during the GUI update.

This allows user to use the command line to close the application instead of pressing on the 'x' button.

Implemented 'bye' command to call a delay function to close the application after a certain time.

Other info
- Fix build.gradle to be able to use gradle to build a jar file.
- Fix message displayed in GUI

JonathanHoshi committed:
Add CI and Remove text-ui-test due to GUI Changes

Remove text-ui-test due to GUI changes
Change yml file to exclude text-ui-test

JonathanHoshi committed:
Add Statistics support

Implement stats feature to allow user to view their task statistics information.

Create a StatsCommand, UserProfile class to store the necessary information of the user and facilitate the displaying of the information in the UI.

Modify the storage system to store user information instead of only task list information.

By extracting the data conversion into their respective class of Task, TaskList, UserProfile, this allows organisation in the code.

Other info
- Update JavaDocs of StatsCommand and missing JavaDocs

JonathanHoshi committed:
Merge branch 'branch-C-Statistics'

JonathanHoshi committed:
Fix storage printing issue

JonathanHoshi committed:
Add BetterGUI

GUI does not visually differentiate error message and bot responses. GUI images is a little too big which result in lesser space for the responses.

Add error message checking which allows the highlight of error messages, turning dialog box to a light pink color. Modify images to be rounded and reduced in size.

This enables user to easily differentiate error messages due to the color differences and allows more information to be displayed at once due to the smaller image.

Other info
- Add window title
- Rename method names (proper naming)

JonathanHoshi committed:
Set theme jekyll-theme-leap-day

JonathanHoshi committed:
Update README v0.1

Update readme to contain todo, deadline, event features and descriptions.

Other info
- Command input to lower case
- Additional formatting for visual display

JonathanHoshi committed:
Update README v0.2

Update readme to contain find, mark, delete, exit, bot, list features and descriptions.

Other info
- Additional formatting for visual display

JonathanHoshi committed:
Update README v1.0

Fix minor formatting issue of readme.

JonathanHoshi committed:
Fix header comment

Minor documentation fix for launcher class.
